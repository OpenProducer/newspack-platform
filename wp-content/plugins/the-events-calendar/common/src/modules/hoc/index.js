export { default as withStore } from './with-store';
export { default as withSaveData } from './with-save-data';
export { default as withForm } from './with-form';
export { default as withBlockCloser } from './with-block-closer';
export { default as withSelected } from './with-selected';
