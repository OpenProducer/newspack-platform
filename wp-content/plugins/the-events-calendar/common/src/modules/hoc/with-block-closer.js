/* eslint-disable max-len */
/**
 * External dependencies
 */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { noop } from 'lodash';

const ESCAPE_KEY = 27;
export const EVENT_NAMESPACE = 'tribe:click:proxy';
export const dispatch = ( e ) => {
	e.target.dispatchEvent( new CustomEvent( EVENT_NAMESPACE, { bubbles: true } ) );
};
export const intercept = e => e.stopPropagation();

export default ( WrappedComponent ) => {
	/**
	 * Prevents clicks on block or blacklisted DOM elements
	 * from closing the block
	 *
	 * @class WithBlockCloser
	 * @extends {PureComponent}
	 */
	class WithBlockCloser extends PureComponent {
		static displayName = `WithBlockCloser( ${ WrappedComponent.displayName || WrappedComponent.name || 'Component ' }`

		static propTypes = {
			onClose: PropTypes.func.isRequired,
			classNameClickBlacklist: PropTypes.arrayOf( PropTypes.string ).isRequired,
			isOpen: PropTypes.bool.isRequired,
		};

		static defaultProps = {
			classNameClickBlacklist: [ '.edit-post-sidebar' ],
			onClose: noop,
			isOpen: false,
		}

		nodeRef = React.createRef();
		_eventNamespace = EVENT_NAMESPACE;

		/**
		 * dispatches custom events
		 *
		 * @memberof WithBlockCloser
		 * @param {Event} e event
		 */
		_dispatchClickProxyEvent = dispatch;

		// Prevent CustomEvents from propagating to document proxy listeners
		_interceptClickProxyEvent = intercept;

		/**
		 * keydown handler
		 *
		 * @memberof WithBlockCloser
		 * @param {Event} e event
		 */
		handleKeyDown = ( e ) => {
			if ( e.keyCode === ESCAPE_KEY ) {
				this.props.onClose();
			}
		}

		handleClick = () => this.props.onClose()

		componentDidMount() {
			this.props.isOpen && this._addEventListeners();
		}

		componentDidUpdate( prevProps ) {
			if ( prevProps.isOpen !== this.props.isOpen ) {
				this.props.isOpen
					? this._addEventListeners()
					: this._removeEventListeners();
			}
		}

		componentWillUnmount() {
			this._removeEventListeners();
		}

		get blacklistedNodes() {
			const classNames = this.props.classNameClickBlacklist.join( ', ' );
			return Array.from( document.querySelectorAll( classNames ) );
		}

		get node() {
			return this.nodeRef.current;
		}

		_addEventListeners() {
			// Intercept custom events bubbled in block or blacklisted nodes
			this.node.addEventListener( this._eventNamespace, this._interceptClickProxyEvent );
			this.blacklistedNodes.forEach(
				node => node.addEventListener( this._eventNamespace, this._interceptClickProxyEvent )
			);

			// Wait to receive custom events, if not intercepted, then go to click handler
			document.addEventListener( this._eventNamespace, this.handleClick );
			// Dispatch custom event on regular clicks
			document.addEventListener( 'click', this._dispatchClickProxyEvent );

			// Close on certain keypresses
			document.addEventListener( 'keydown', this.handleKeyDown );
		}

		_removeEventListeners() {
			this.node.removeEventListener( this._eventNamespace, this._interceptClickProxyEvent );
			this.blacklistedNodes.forEach(
				node => node.removeEventListener( this._eventNamespace, this._interceptClickProxyEvent )
			);

			document.removeEventListener( 'keydown', this.handleKeyDown );
			document.removeEventListener( this._eventNamespace, this.handleClick );
			document.removeEventListener( 'click', this._dispatchClickProxyEvent );
		}

		render() {
			return (
				<div ref={ this.nodeRef }>
					<WrappedComponent { ...this.props } />
				</div>
			);
		}
	}

	return WithBlockCloser;
};
