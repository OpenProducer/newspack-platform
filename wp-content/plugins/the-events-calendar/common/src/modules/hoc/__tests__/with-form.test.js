/**
 * External dependencies
 */
import renderer from 'react-test-renderer';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

/**
 * Internal dependencies
 */
import { withForm } from '@moderntribe/common/hoc';

const initialState = {
	events: {
	},
	forms: {
		byId: {},
	},
};
// here it is possible to pass in any middleware if needed into //configureStore
const mockStore = configureStore( [ thunk ] );
const store = mockStore( initialState );

const Block = () => <div>With Form!</div>;
let setFormID;
let Wrapper;
let component;
let instance;

describe( 'HOC - With Form', () => {
	beforeEach( () => {
		setFormID = jest.fn( () => 'posts' );
		Wrapper = withForm( setFormID )( Block );
		component = renderer.create( <Wrapper store={ store } postType="post"/> );
		instance = component.root;
	} );

	afterEach( () => {
		mockStore( initialState );
		store.clearActions();
		setFormID.mockClear();
	} );

	it( 'Should render a component', () => {
		expect( component.toJSON() ).toMatchSnapshot();
	} );

	it( 'Should render the inner component', () => {
		expect( instance ).not.toBe( null );
		expect( () => instance.findByType( Block ) ).not.toThrowError();
	} );

	it( 'Should attach the form properties', () => {
		const expected = {
			edit: false,
			create: false,
			fields: {},
			submit: false,
		};
		expect( instance.findByType( Block ).props ).toMatchObject( expected );
	} );

	it( 'Should have properties as functions', () => {
		const props = instance.findByType( Block ).props;
		const expectedProps = [
			'maybeRemoveEntry',
			'setSubmit',
			'sendForm',
			'editEntry',
			'createDraft',
		];

		expectedProps.forEach( ( property ) => {
			expect( props[ property ] ).not.toBeUndefined();
			expect( typeof props[ property ] ).toBe( 'function' );
		} );
	} );

	it( 'Should register the postType by dispatching the actions', () => {
		expect( store.getActions() ).toMatchSnapshot();
	} );

	it( 'Should register the ID of the form', () => {
		expect( setFormID ).toHaveBeenCalled();
		expect( setFormID ).toHaveBeenCalledTimes( 3 );
	} );
} );

