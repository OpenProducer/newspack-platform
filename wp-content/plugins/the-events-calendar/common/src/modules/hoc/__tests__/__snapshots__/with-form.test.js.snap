// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC - With Form Should register the postType by dispatching the actions 1`] = `
Array [
  Object {
    "payload": Object {
      "id": "posts",
      "type": "post",
    },
    "type": "@@MT/COMMON/ADD_FORM",
  },
]
`;

exports[`HOC - With Form Should render a component 1`] = `
<div>
  With Form!
</div>
`;
